src/hockeypuck/pghkp/storage.go

```diff
@@ -463,50 +661,431 @@ func (st *storage) insertKeyTx(tx *sql.Tx, key *openpgp.PrimaryKey) (isDuplicate
+	txStrs := []string{bulkTxInsertKeys, bulkTxInsertSubkeys}
+	msgStrs := []string{"bulkTx-insert-keys", "bulkTx-insert-subkeys"}
+	err := st.bulkInsertSingleTx(txStrs, msgStrs)
+	if err != nil {
+		result.Errors = append(result.Errors, err)
+		return 0, 0, false
+	}
+	return nullKeys, nullSubkeys, true
+}
+
+func (st *storage) bulkInsertSendBunchTx(keystmt, msgSpec string, keysValueArgs []interface{}) (err error) {
+	// In single transaction...
+	tx, err := st.Begin()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer func() {
+		if err != nil {
+			tx.Rollback()
+		} else {
+			tx.Commit()
 		}
+	}()
+
+	stmt, err := tx.Prepare(keystmt)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	_, err = stmt.Exec(keysValueArgs...) // All keys in bunch
+	if err != nil {
+		return errors.Wrapf(err, "cannot simply send a bunch of %s to server (too large bunch?)", msgSpec)
+	}
+	err = stmt.Close()
+	if err != nil {
+		return errors.Wrapf(err, "failed to close xfer sending a bunch of %s to server", msgSpec)
+	}
+
+	return nil
+}
 
-		if isDuplicate, err := st.insertKey(key); err != nil {
+func (st *storage) bulkInsertSendBunch(keystmt, subkeystmt string, keysValueArgs, subkeysValueArgs []interface{}) (err error) {
+
+	// Send all keys to in-mem tables to the pg server; *no constraints checked*
+	err = st.bulkInsertSendBunchTx(keystmt, "keys", keysValueArgs)
+	if err != nil {
+		return err
+	}
+
+	// Send all subkeys to in-mem tables to the pg server; *no constraints checked*
+	err = st.bulkInsertSendBunchTx(subkeystmt, "subkeys", subkeysValueArgs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type KeyInsertArgs struct {
+	RFingerprint *string
+	jsonStrDoc   *string
+	MD5          *string
+	keywords     *string
+}
+type SubkeyInsertArgs struct {
+	keyRFingerprint    *string
+	subkeyRFingerprint *string
+}
+
+// Insert keys & subkeys to in-mem tables with no constraints at all: should have no errors!
+func (st *storage) bulkInsertDoCopy(keyInsertArgs []KeyInsertArgs, subkeyInsertArgs [][]SubkeyInsertArgs,
+	result *hkpstorage.InsertError) (ok bool) {
+	lenKIA := len(keyInsertArgs)
+	for totKeyArgs, totSubkeyArgs, idx, lastIdx := 0, 0, 0, 0; idx < lenKIA; totKeyArgs, totSubkeyArgs, lastIdx = 0, 0, idx {
+		keysValueStrings := make([]string, 0, 5000)         // could be 10K keys but 5K more uniform for 15K files
+		keysValueArgs := make([]interface{}, 0, 5000*6)     // *** less than 64k arguments ***
+		subkeysValueStrings := make([]string, 0, 32000)     // max 32K subkeys
+		subkeysValueArgs := make([]interface{}, 0, 32000*2) // *** less than 64k arguments ***
+		insTime := make([]time.Time, 0, 5000)               // stupid but anyway...
+		for i, j := 0, 0; idx < lenKIA; idx, i = idx+1, i+1 {
+			lenSKIA := len(subkeyInsertArgs[idx])
+			totKeyArgs += 6
+			totSubkeyArgs += 2 * lenSKIA
+			if (totKeyArgs > 30000) || (totSubkeyArgs > 64000) {
```

Comment:
I'd prefer these numbers were consts with a nice godoc description of why... sounds like some of these are upper limits on what PostgreSQL allows in SQL statements?

Reply:
Yes, you're right, of course; those should be consts. I'll make them.

Actually, I just read somewhere (on the internet) that 64K arguments is the maximum (did not test it). I will look if I can find the exact docs...

Reply:
[This stackoverflow question](https://stackoverflow.com/questions/6581573/what-are-the-max-number-of-allowable-parameters-per-database-provider-type/49379324#49379324) deals with max parameters. One answer for PostreSQL saays:
"The PostgreSQL wire protocol uses 16-bit integers for count of parameters in the bind message (http://www.postgresql.org/docs/9.4/static/protocol-message-formats.html). "
([Current version protocol spec link](https://www.postgresql.org/docs/current/protocol-message-formats.html))

So, unless there is some additional limit, somewhere (I have not found any), this must be it: up to 65535 parameters. Someone on that thread says has tested it and works up to 65535 parameters, but not 65536.

Reply:
Did the comments and also updated the comments of `bulkIns...` queries for godoc.